File: src/gitscale/commands/remove.py
"""Remove a sub-repository entry from .gitscale config."""

from pathlib import Path

import click

from gitscale.config import (
    find_config,
    load_config,
    write_config,
)


@click.command()
@click.argument("directory")
@click.option(
    "-C",
    "--root",
    type=click.Path(exists=True, path_type=Path),
    default=None,
    help="Root directory containing .gitscale.toml (default: auto-detect).",
)
@click.pass_context
def remove(
    ctx: click.Context,
    directory: str,
    root: Path | None,
) -> None:
    """Remove a sub-repository entry from .gitscale config.

    DIRECTORY is the local subdirectory name to remove.
    """
    config_path = find_config(root)
    config = load_config(config_path)

    remaining = [e for e in config.repos if e.directory != directory]

    if len(remaining) == len(config.repos):
        raise click.ClickException(
            f"'{directory}' is not declared in {config_path}"
        )

    write_config(config_path, remaining, storage_url=config.storage_url)
    click.echo(f"Removed {directory}")

File: src/commands/remove.rs
use anyhow::Result;
use std::io::Write;
use std::path::Path;

use crate::config::{find_config, load_config, write_config};

pub fn run(directory: &str, root: Option<&Path>, out: &mut dyn Write) -> Result<()> {
    let config_path = find_config(root)?;
    let mut config = load_config(&config_path)?;
    let orig_len = config.repos.len();

    config.repos.retain(|e| e.directory != directory);

    if config.repos.len() == orig_len {
        anyhow::bail!(
            "'{}' is not declared in {}",
            directory,
            config_path.display()
        );
    }

    write_config(&config_path, &config)?;
    writeln!(out, "Removed {}", directory)?;
    Ok(())
}
